torch/csrc/jit/codegen/cuda/ir_internal_nodes.h

```diff
@@ -302,6 +303,241 @@ struct TORCH_CUDA_API Reorder : public Expr {
   const std::vector<int> pos2axis_;
 };
 
+/*
+ * ForLoop provides scoping around an int iterator from 0 to range. Exprs placed
+ * in its body are considered inside the scope of the for loop. In the future
+ * the implementation should look quite different so that we can do proper
+ * dependency annalysis like in Fusion.
+ *
+ * TODO: Change implmentation of Exprs contained in the scope to be more similar
+ * to Fusion where we can do proper dependency analysis.
+ */
+struct TORCH_API ForLoop : public Expr {
+  ~ForLoop() = default;
+  ForLoop(
+      Val* _index,
+      IterDomain* _range,
+      const std::vector<Expr*>& _body = {},
+      Expr* parent_scope = nullptr);
+
+  ForLoop(const ForLoop& other) = delete;
+  ForLoop& operator=(const ForLoop& other) = delete;
+
+  ForLoop(ForLoop&& other) = delete;
+  ForLoop& operator=(ForLoop&& other) = delete;
+
+  Val* index() const noexcept {
+    return index_;
+  }
+
+  IterDomain* range() const noexcept {
+    return range_;
+  }
+
+  Scope& body() noexcept {
+    return body_;
+  }
+
+  const Scope& constBody() const noexcept {
```

Comment:
Is it necessary to separate `body()` with `constBody()`? can't we do this with `const Scope& body() const noexcept`?
This is more of a question than a comment :)

Reply:
Same for ifThenElse

Reply:
You can't overload functions based on return type. Sometimes we explicitly need a const body reference. Primarily because we need to access body in other ForLoop/IfThenElse member functions that are const. For example SameAs, the printer (also takes const args), however, in other instances like code lowering we want to be able to modify the body contents. We could make body() protected and use friend relationships, but that seemed excessive to me and I like to avoid friend relationships unless we're protecting user interfaces from modifying things they shouldn't (like TenosrView).

Reply:
We are not overloading on the return type, but on `const correctness` for the function instead (that `const` after `body()`)
So the overloading should work fine in cases where we have to abide to const qualifier. I searched the PR and all places using `constBody` are all marked properly with const qualifier either in the function or pointer already. 

It will be a little different on the API. Now when you want 
`ForLoop loop(...); const auto& body = x.body()` we use non-const accessor but attach the const qualifier to the variable. This looks cleaner in my opinion.